### PR TITLE
fix(ci): include e2e tests in coverage

### LINKED ISSUE #92: [Feature]: Include e2e tests in coverage
### What is the feature you would like to see?

We don't currently include e2e tests in coverage because some setup is required to run them. Specifically, you need `solc` and `nitro-testnode` to be available.

We should either create a new workflow that uploads this to codecov or update the current coverage workflow to include e2e tests.

### Contribution Guidelines

- [X] I agree to follow this project's Contribution Guidelines

### PR DESCRIPTION
**Pull request by alexfertel**
Resolves #92

### CODE REVIEW

**Review by bidzyyys — APPROVED** [review]
looks good!

File: lib/e2e/src/project.rs [hunk 1]
```diff
@@ -49,13 +51,32 @@ fn get_wasm(name: &str) -> eyre::Result<PathBuf> {
...
-    let executable = env::current_exe()?;
-    let out_dir = executable
-        .parent()
-        .expect("executable path should have a final component")
-        .join("../../");
-    let wasm: PathBuf =
-        out_dir.join(format!("wasm32-unknown-unknown/release/{name}.wasm"));
+    let mut target_dir = env::current_exe()?;
+
+    // Recursively find a `target` directory.
```
Comment (qalisander) [inline lib/e2e/src/project.rs:56]: I would say this approach is much better for distributing library. But a bit hacky:)

File: .github/workflows/e2e-tests.yml [hunk 2]
```diff
@@ -63,3 +63,19 @@ jobs:
         run: |
           export NIGHTLY_TOOLCHAIN=${{steps.toolchain.outputs.name}}
           ./scripts/e2e-tests.sh
+      # Coverage
+      - name: cargo install cargo-llvm-cov
```
Comment (qalisander) [inline .github/workflows/e2e-tests.yml:67]: Can you explain how it should work?
Are we checking test coverage of test framework here or integration tests?
And if it is last one how it works?

**Review by qalisander — COMMENTED** [review]
Great!

File: lib/e2e/src/project.rs (hunk 1 above)
Comment (alexfertel) [reply lib/e2e/src/project.rs:56]: Yeah, we should probably find a better way in the future, I'm not that happy with it. Nonetheless, it should be fine for most use cases.

File: .github/workflows/e2e-tests.yml (hunk 2 above)
Comment (alexfertel) [reply .github/workflows/e2e-tests.yml:67]: Ofc! So, this is running the `e2e` tests and collecting coverage data, which is uploaded to codecov.io and merged with the data uploaded from the `test` workflow.

> Are we checking test coverage of test framework here, unit tests code or contracts library?

Not entirely sure what you mean. We have two workflows right now that collect coverage data: `test` and this one (`e2e`).

* `test` runs `cargo test --features std`
* `e2e` runs `cargo test --features std,e2e --test '*'`

So, the first one runs unit tests, and the second one runs integration tests.

Does this answer your questions?

Comment (qalisander) [reply .github/workflows/e2e-tests.yml:67]: I mean it in a way of how code coverage implemented inside. Is it changing binary code to insert counters or smth else? And will it work for wasm binaries deployed to nitro node?

Comment (alexfertel) [reply .github/workflows/e2e-tests.yml:67]: ah, no clue hehe

Comment (alexfertel) [reply .github/workflows/e2e-tests.yml:67]: I expect it to work, yes. for example, events are getting covered, but we need to write more tests and see how the diff changes.

Comment (qalisander) [reply .github/workflows/e2e-tests.yml:67]: oke. Will check out then

**Review by qalisander — APPROVED** [review]

**Comment by qalisander** [comment]
Also seems that test coverage adds few minutes to pipeline to pass

**Comment by alexfertel** [comment]
> Also seems that test coverage adds few minutes to pipeline to pass

Yeah, coverage has to be a separate workflow, let me fix that.

**Pull request merged**